File: app/src/main/java/me/zhanghai/android/files/compat/ProxyFileDescriptorCallbackCompat.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.compat;

import android.os.Build;
import android.os.ProxyFileDescriptorCallback;
import android.system.ErrnoException;
import android.system.OsConstants;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

/**
 * @see ProxyFileDescriptorCallback
 */
public abstract class ProxyFileDescriptorCallbackCompat {

    /**
     * @see ProxyFileDescriptorCallback#onGetSize()
     */
    public long onGetSize() throws ErrnoException {
        throw new ErrnoException("onGetSize", OsConstants.EBADF);
    }

    /**
     * @see ProxyFileDescriptorCallback#onRead(long, int, byte[])
     */
    public int onRead(long offset, int size, byte[] data) throws ErrnoException {
        throw new ErrnoException("onRead", OsConstants.EBADF);
    }

    /**
     * @see ProxyFileDescriptorCallback#onWrite(long, int, byte[])
     */
    public int onWrite(long offset, int size, byte[] data) throws ErrnoException {
        throw new ErrnoException("onWrite", OsConstants.EBADF);
    }

    /**
     * @see ProxyFileDescriptorCallback#onFsync()
     */
    public void onFsync() throws ErrnoException {
        throw new ErrnoException("onFsync", OsConstants.EINVAL);
    }

    /**
     * @see ProxyFileDescriptorCallback#onRelease()
     */
    abstract public void onRelease();

    @NonNull
    @RequiresApi(Build.VERSION_CODES.O)
    ProxyFileDescriptorCallback toProxyFileDescriptorCallback() {
        return new ProxyFileDescriptorCallback() {
            @Override
            public long onGetSize() throws ErrnoException {
                return ProxyFileDescriptorCallbackCompat.this.onGetSize();
            }
            @Override
            public int onRead(long offset, int size, byte[] data) throws ErrnoException {
                return ProxyFileDescriptorCallbackCompat.this.onRead(offset, size, data);
            }
            @Override
            public int onWrite(long offset, int size, byte[] data) throws ErrnoException {
                return ProxyFileDescriptorCallbackCompat.this.onWrite(offset, size, data);
            }
            @Override
            public void onFsync() throws ErrnoException {
                ProxyFileDescriptorCallbackCompat.this.onFsync();
            }
            @Override
            public void onRelease() {
                ProxyFileDescriptorCallbackCompat.this.onRelease();
            }
        };
    }
}

File: app/src/main/java/me/zhanghai/android/files/compat/ProxyFileDescriptorCallbackCompat.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.compat

import android.os.Build
import android.os.ProxyFileDescriptorCallback
import android.system.ErrnoException
import android.system.OsConstants
import androidx.annotation.RequiresApi

abstract class ProxyFileDescriptorCallbackCompat {
    @Throws(ErrnoException::class)
    open fun onGetSize(): Long {
        throw ErrnoException("onGetSize", OsConstants.EBADF)
    }

    @Throws(ErrnoException::class)
    open fun onRead(offset: Long, size: Int, data: ByteArray): Int {
        throw ErrnoException("onRead", OsConstants.EBADF)
    }

    @Throws(ErrnoException::class)
    open fun onWrite(offset: Long, size: Int, data: ByteArray): Int {
        throw ErrnoException("onWrite", OsConstants.EBADF)
    }

    @Throws(ErrnoException::class)
    open fun onFsync() {
        throw ErrnoException("onFsync", OsConstants.EINVAL)
    }

    abstract fun onRelease()

    @RequiresApi(Build.VERSION_CODES.O)
    fun toProxyFileDescriptorCallback(): ProxyFileDescriptorCallback {
        return object : ProxyFileDescriptorCallback() {
            @Throws(ErrnoException::class)
            override fun onGetSize(): Long = this@ProxyFileDescriptorCallbackCompat.onGetSize()

            @Throws(ErrnoException::class)
            override fun onRead(offset: Long, size: Int, data: ByteArray): Int =
                this@ProxyFileDescriptorCallbackCompat.onRead(offset, size, data)

            @Throws(ErrnoException::class)
            override fun onWrite(offset: Long, size: Int, data: ByteArray): Int =
                this@ProxyFileDescriptorCallbackCompat.onWrite(offset, size, data)

            @Throws(ErrnoException::class)
            override fun onFsync() = this@ProxyFileDescriptorCallbackCompat.onFsync()

            override fun onRelease() = this@ProxyFileDescriptorCallbackCompat.onRelease()
        }
    }
}
